File: python/src/vtx_recon/providers/gitlab.py
"""GitLab capability ladder — prove depth of access for a leaked PAT.

Handles TruffleHog ``GitLab`` findings. A GitLab personal access token
(``glpat-...``) authenticates with the ``PRIVATE-TOKEN`` header. Two SAFE rungs:

  1. ``gitlab.user``         ``GET /api/v4/user`` — confirms identity. Decides
     VALID vs DENIED. Read-only, idempotent.
  2. ``gitlab.token.scopes`` ``GET /api/v4/personal_access_tokens/self`` —
     reveals the token's exact scopes (depth of access) without exercising any
     of them. Read-only.

Every rung is ordered (identity first, then depth), READ-ONLY, and the ladder
never raises across the public boundary: failures become a :class:`ProbeResult`
with ``success=False`` so one dead key cannot crash a batch run. The raw token
is held only transiently for the HTTP call and never lands in
:attr:`ProbeResult.evidence`.

AUTHORIZED USE ONLY. See TERMS.md.
"""

from __future__ import annotations

import httpx

from ..models import Finding, LadderResult, ProbeResult, Verdict
from ..safety import Consent, ProbeTier
from . import register

__all__ = ["DETECTORS", "gitlab_ladder"]

# TruffleHog DetectorName values routed to this ladder (matched case-insensitively).
DETECTORS = ("GitLab",)

API_BASE = "https://gitlab.com/api/v4"

# A single shared timeout: probes must be quick and must never hang a batch.
_TIMEOUT = httpx.Timeout(10.0)


def _network_failure(name: str, tier: ProbeTier, exc: Exception) -> ProbeResult:
    """Turn an httpx/transport error into a non-success rung (never raise)."""
    return ProbeResult(
        name=name,
        tier=tier,
        success=False,
        detail=f"probe could not complete: {type(exc).__name__}",
        evidence={"error": type(exc).__name__},
    )


def _verdict_from(rungs: list[ProbeResult]) -> Verdict:
    """Derive the impact tier from the rungs that ran.

    * A successful GATED rung that actually ran (not blocked) -> PROVEN.
    * Any successful SAFE rung -> VALID (authenticates + depth shown).
    * The key authenticated nowhere -> DENIED.
    """
    if any(r.success and r.tier is ProbeTier.GATED and not r.blocked for r in rungs):
        return Verdict.PROVEN
    if any(r.success for r in rungs):
        return Verdict.VALID
    return Verdict.DENIED


@register("GitLab")
async def gitlab_ladder(finding: Finding, consent: Consent) -> LadderResult:
    """GitLab ladder: SAFE identity (/user) -> SAFE token scopes (self)."""
    scope = consent.require_ladder_scope()
    rungs: list[ProbeResult] = []
    token = finding.raw

    identity = await _gitlab_user(token)
    rungs.append(identity)
    # Only probe depth if the token authenticated at all (ordered ladder).
    if identity.success:
        rungs.append(await _gitlab_token_scopes(token))

    return LadderResult(
        finding=finding,
        provider="gitlab",
        verdict=_verdict_from(rungs),
        rungs=rungs,
        authorized_scope=scope,
    )


async def _gitlab_user(token: str) -> ProbeResult:
    """SAFE: ``GET /api/v4/user`` confirms identity."""
    name = "gitlab.user"
    try:
        async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
            resp = await client.get(f"{API_BASE}/user", headers={"PRIVATE-TOKEN": token})
    except httpx.HTTPError as exc:
        return _network_failure(name, ProbeTier.SAFE, exc)

    if resp.status_code != 200:
        return ProbeResult(
            name=name,
            tier=ProbeTier.SAFE,
            success=False,
            detail=f"token rejected (HTTP {resp.status_code})",
            evidence={"status": resp.status_code},
        )

    try:
        body = resp.json()
    except ValueError as exc:
        return _network_failure(name, ProbeTier.SAFE, exc)

    return ProbeResult(
        name=name,
        tier=ProbeTier.SAFE,
        success=True,
        detail=f"authenticated as {body.get('username')} (id {body.get('id')})",
        evidence={
            "status": resp.status_code,
            "id": body.get("id"),
            "username": body.get("username"),
            "is_admin": body.get("is_admin"),
        },
    )


async def _gitlab_token_scopes(token: str) -> ProbeResult:
    """SAFE: ``GET /api/v4/personal_access_tokens/self`` reveals scopes.

    Reading the token's own scopes proves depth of access without exercising
    any of them.
    """
    name = "gitlab.token.scopes"
    try:
        async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
            resp = await client.get(
                f"{API_BASE}/personal_access_tokens/self",
                headers={"PRIVATE-TOKEN": token},
            )
    except httpx.HTTPError as exc:
        return _network_failure(name, ProbeTier.SAFE, exc)

    if resp.status_code != 200:
        return ProbeResult(
            name=name,
            tier=ProbeTier.SAFE,
            success=False,
            detail=f"could not read token scopes (HTTP {resp.status_code})",
            evidence={"status": resp.status_code},
        )

    try:
        body = resp.json()
    except ValueError as exc:
        return _network_failure(name, ProbeTier.SAFE, exc)

    scopes = body.get("scopes") or []
    return ProbeResult(
        name=name,
        tier=ProbeTier.SAFE,
        success=True,
        detail=f"token scopes: {', '.join(scopes) if scopes else '(none)'}",
        evidence={
            "status": resp.status_code,
            "scopes": scopes,
            "active": body.get("active"),
        },
    )

File: node/src/providers/gitlab.ts
/**
 * GitLab capability ladder — prove depth of access for a leaked PAT.
 *
 * Handles TruffleHog `GitLab` findings. A GitLab personal access token
 * (`glpat-...`) authenticates with the `PRIVATE-TOKEN` header. Two SAFE rungs:
 *
 *   1. `gitlab.user`         `GET /api/v4/user` — confirms identity. Decides
 *      VALID vs DENIED. Read-only, idempotent.
 *   2. `gitlab.token.scopes` `GET /api/v4/personal_access_tokens/self` — reveals
 *      the token's exact scopes (depth of access) without exercising any of
 *      them. Read-only.
 *
 * Every rung is ordered (identity first, then depth), READ-ONLY, and the ladder
 * never throws across its public boundary: failures become a {@link ProbeResult}
 * with `success=false`. The raw token is held only transiently for the HTTP call
 * and never lands in evidence.
 *
 * AUTHORIZED USE ONLY. See TERMS.md.
 */

import { Finding, LadderResult, ProbeResult, Verdict } from "../models.js";
import { Consent, ProbeTier } from "../safety.js";
import { HttpError, httpRequest, readJson, type FetchLike } from "./http.js";
import { register } from "./registry.js";

/** TruffleHog DetectorName values routed to this ladder (case-insensitive). */
export const DETECTORS = ["GitLab"] as const;

const API_BASE = "https://gitlab.com/api/v4";

// A single shared timeout: probes must be quick and must never hang a batch.
const TIMEOUT_MS = 10_000;

function networkFailure(name: string, tier: ProbeTier, exc: unknown): ProbeResult {
  return new ProbeResult({
    name,
    tier,
    success: false,
    detail: `probe could not complete: ${errName(exc)}`,
    evidence: { error: errName(exc) },
  });
}

function errName(exc: unknown): string {
  if (exc instanceof HttpError) return exc.kind;
  return exc instanceof Error ? exc.constructor.name : "Error";
}

/**
 * Derive the impact tier from the rungs that ran.
 * - A successful GATED rung that actually ran (not blocked) -> PROVEN.
 * - Any successful SAFE rung -> VALID (authenticates + depth shown).
 * - The key authenticated nowhere -> DENIED.
 */
function verdictFrom(rungs: ProbeResult[]): Verdict {
  if (rungs.some((r) => r.success && r.tier === ProbeTier.GATED && !r.blocked)) {
    return Verdict.PROVEN;
  }
  if (rungs.some((r) => r.success)) {
    return Verdict.VALID;
  }
  return Verdict.DENIED;
}

/** GitLab ladder: SAFE identity (/user) -> SAFE token scopes (self). */
export async function gitlabLadder(
  finding: Finding,
  consent: Consent,
  options: { fetchImpl?: FetchLike } = {},
): Promise<LadderResult> {
  const scope = consent.requireLadderScope();
  const rungs: ProbeResult[] = [];
  const token = finding.raw;

  const identity = await gitlabUser(token, options.fetchImpl);
  rungs.push(identity);
  // Only probe depth if the token authenticated at all (ordered ladder).
  if (identity.success) {
    rungs.push(await gitlabTokenScopes(token, options.fetchImpl));
  }

  return new LadderResult({
    finding,
    provider: "gitlab",
    verdict: verdictFrom(rungs),
    rungs,
    authorizedScope: scope,
  });
}

/** SAFE: `GET /api/v4/user` confirms identity. */
async function gitlabUser(token: string, fetchImpl: FetchLike | undefined): Promise<ProbeResult> {
  const name = "gitlab.user";
  let resp: Response;
  try {
    resp = await httpRequest(`${API_BASE}/user`, {
      headers: { "PRIVATE-TOKEN": token },
      timeoutMs: TIMEOUT_MS,
      fetchImpl,
    });
  } catch (exc) {
    return networkFailure(name, ProbeTier.SAFE, exc);
  }

  if (resp.status !== 200) {
    return new ProbeResult({
      name,
      tier: ProbeTier.SAFE,
      success: false,
      detail: `token rejected (HTTP ${resp.status})`,
      evidence: { status: resp.status },
    });
  }

  const body = (await readJson(resp)) as Record<string, unknown> | undefined;
  if (body === undefined) {
    return networkFailure(name, ProbeTier.SAFE, new SyntaxError("invalid JSON"));
  }

  return new ProbeResult({
    name,
    tier: ProbeTier.SAFE,
    success: true,
    detail: `authenticated as ${body["username"]} (id ${body["id"]})`,
    evidence: {
      status: resp.status,
      id: body["id"] ?? null,
      username: body["username"] ?? null,
      is_admin: body["is_admin"] ?? null,
    },
  });
}

/** SAFE: `GET /api/v4/personal_access_tokens/self` reveals scopes. */
async function gitlabTokenScopes(token: string, fetchImpl: FetchLike | undefined): Promise<ProbeResult> {
  const name = "gitlab.token.scopes";
  let resp: Response;
  try {
    resp = await httpRequest(`${API_BASE}/personal_access_tokens/self`, {
      headers: { "PRIVATE-TOKEN": token },
      timeoutMs: TIMEOUT_MS,
      fetchImpl,
    });
  } catch (exc) {
    return networkFailure(name, ProbeTier.SAFE, exc);
  }

  if (resp.status !== 200) {
    return new ProbeResult({
      name,
      tier: ProbeTier.SAFE,
      success: false,
      detail: `could not read token scopes (HTTP ${resp.status})`,
      evidence: { status: resp.status },
    });
  }

  const body = (await readJson(resp)) as Record<string, unknown> | undefined;
  if (body === undefined) {
    return networkFailure(name, ProbeTier.SAFE, new SyntaxError("invalid JSON"));
  }

  const scopes = (body["scopes"] as string[] | undefined) ?? [];
  return new ProbeResult({
    name,
    tier: ProbeTier.SAFE,
    success: true,
    detail: `token scopes: ${scopes.length > 0 ? scopes.join(", ") : "(none)"}`,
    evidence: { status: resp.status, scopes, active: body["active"] ?? null },
  });
}

register([...DETECTORS], (finding, consent) => gitlabLadder(finding, consent));
